Implement genProfilingEnterCallback genProfilingLeaveCallback on Arm64

echesakovMSFT opened the pull request:
Fixes https://github.com/dotnet/coreclr/issues/19368

The following conventions for `genProfilingEnterCallback` and `genProfilingLeaveCallback` will be used on arm64:
- **x10** is used to pass `functionIDOrClientID` to a helper (`REG_PROFILER_ENTER_ARG_FUNC_ID` or `REG_PROFILER_LEAVE_ARG_FUNC_ID`)
- **x11** is used to pass `callerSp` to a helper (`REG_PROFILER_ENTER_ARG_CALLER_SP` or `REG_PROFILER_LEAVE_ARG_CALLER_SP`). It's quite tricky to figure out the callerSp on arm64 since we have 5 or 6 different types of the stack frame and fp is not guaranteed to point at the top of the stack frame (as on x64 or arm). 
- **x12** is used for call target (i.e. an address of a helper function or jump stub) (`REG_DEFAULT_HELPER_CALL_TARGET`)
contains call target (i.e. address of the helper function or jump stub)
- During the call the following volatile registers are NOT preserved: x9-x17, v16-v31
- During the call to `CORINFO_HELP_PROF_FCN_ENTER` all the argument registers (x0-x7, x8, v0-v7) will contain valid values of the arguments and must be preserved during the call. This helps to avoid pre-spilling of the argument register as we do on arm32.

echesakovMSFT commented:
The following tests are failing:
- JIT/Directed/VectorABI/VectorMgdMgd_ro/VectorMgdMgd_ro.sh
- JIT/Directed/VectorABI/VectorMgdMgdStatic_r/VectorMgdMgdStatic_r.sh
- JIT/Directed/VectorABI/VectorMgdMgdArray_r/VectorMgdMgdArray_r.sh
- JIT/Directed/VectorABI/VectorMgdMgdArray_ro/VectorMgdMgdArray_ro.sh
- JIT/Directed/VectorABI/VectorMgdMgd_r/VectorMgdMgd_r.sh
- JIT/Directed/VectorABI/VectorMgdMgdStatic_ro/VectorMgdMgdStatic_ro.sh

due to an assertion in src/jit/gentree.cpp:14568

echesakovMSFT commented:
@davmason @noahfalk Do you have any suggestions regarding what ABI should be chosen for CORINFO_HELP_PROF_FCN_ENTER, CORINFO_HELP_PROF_FCN_LEAVE, CORINFO_HELP_PROF_FCN_TAILCALL helpers on Arm64? 

I proposed and implemented the calling convention as described in https://github.com/dotnet/coreclr/pull/26460#issue-312916974 since I could not find any documentation that would prescribe what ABI should be used. 

The implementation passes Pri1 tests except the few tests as mentioned in the previous post. That issue will be fixed separately since it reproduces not only during profiling hooks code generation.

cc @dotnet/jit-contrib

davmason commented:
@echesakovMSFT What you propose seems reasonable, although I am far from an expert on arm64. I think it is important that we pick a calling convention and document it, as long as it is reasonably performant and allows us to do both fast and slow path ELT the exact ABI is flexible.

noahfalk commented:
I agree with @davmason, as long as the ABI is something the JIT can generate reasonable efficiently and we've documented it, the exact choice seems less critical. I am not remotely an ARM64 expert but what you described above seems reasonable to me. If it helps there is an old PR that I never finished that was trying to make some headway documenting some of our pre-existing ELT ABI work: https://github.com/dotnet/coreclr/pull/19560. I don't recall where I got the ARM64 part of that doc so don't feel like you need to give it any weight. 

I do strongly recommend that we document the ABI in the clr-abi doc. I know we haven't been good at that in the past but it shouldn't be hard to avoid making the hole any deeper while the info is fresh : )

echesakovMSFT commented:
I will still need to implement Windows/arm64 helpers but I suspect they would look almost like the Linux ones

@davmason Can you please take a look at the profiler related changes?

BruceForstall commented on src/jit/codegenarm64.cpp:
```diff
@@ -5883,6 +5883,99 @@ void CodeGen::genHWIntrinsicShaRotateOp(GenTreeHWIntrinsic* node)
+
+    gcInfo.gcMarkRegSetNpt(RBM_PROFILER_LEAVE_ARG_FUNC_ID);
+
+    if (compiler->lvaDoneFrameLayout == Compiler::FINAL_FRAME_LAYOUT)
```
Couldn't this just be an assert? We shouldn't ever call this function before final frame layout, should we?

echesakovMSFT replied:
This is true for genProfilingEnterCallback - I am not sure about genProfilingLeaveCallback

echesakovMSFT replied:
After discussion with Bruce we think that the pattern
```c++
if (compiler->lvaDoneFrameLayout == Compiler::FINAL_FRAME_LAYOUT)
{
}
else
{
}
```
in genProfilingLeaveCallback was used because of legacy backend.

It looks that we can remove the else-logic and just do an assertion. I will follow up in another PR and remove the logic on x64 and arm64.

BruceForstall commented on src/vm/arm/asmhelpers.S:
```diff
@@ -561,6 +561,15 @@ ThePreStubPatchLabel:
+// EXTERN_C void JIT_ProfilerEnterLeaveTailcallStub(UINT_PTR ProfilerHandle)
+//
+LEAF_ENTRY JIT_ProfilerEnterLeaveTailcallStub, _TEXT
+    bx lr
```
Is this a temporary/placeholder thing? Or is this ever used in a full implementation?

davmason replied:
It's used if the profiler wants to only have some callbacks and not others (i.e. if they just want enter callbacks then we will register the taillcall stub and leave stub that just do rets for the other two).

echesakovMSFT commented:
/azp run coreclr-outerloop

echesakovMSFT commented:
**More changes**:

1. Fixed an issue in codegenarm.cpp where helper argument in genProfilingLeaveCallback was ignored
2. Refactored the code in genProfilingLeaveCallback in codegenarm.cpp to use genTransferRegState
3. Removed some unused macro in target.h
4. As suggested removed the non-FINAL_FRAME_LAYOUT for arm64 - we don't need an assert since lvaToCallerSPRelativeOffset has one

echesakovMSFT commented:
I have update the first post to indicate what ABI is chosen for the ELT hooks.

echesakovMSFT commented:
I am going to follow-up on the @noahfalk's PR  #19560 and update the information for Arm32 and Arm64
